### PR TITLE
Sklearn adaptation

### PR DESCRIPTION
**Pull request by rosecers**
These are the versions of `kpcovr` and `pcovr` that @bhelfrecht  and I wrote using _as best as we could gather_ sklearn-syntax. It's important that we merge them soon to allow @hurrican642 to build off of what's already been written. Discuss!

### CODE REVIEW

File: utilities/sklearn_covr/kpcovr.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+        See :term:`Glossary <random_state>`.
+
+        .. versionadded:: 0.18
+
+    copy_X : boolean, default=True
+        If True, input X is copied and stored by the model in the `X_fit_`
+        attribute. If no further changes will be done to X, setting
+        `copy_X=False` saves memory by storing a reference.
+
+        .. versionadded:: 0.18
```
Comment (Luthaf) [inline utilities/sklearn_covr/kpcovr.py:83]: is that left-over from copy-pasting?

File: tests/pcovr.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    pcovr.fit(X, Y)
+    T = pcovr.transform(X)
+
+    assert check_X_y(X, T, multi_output=True)
+
+    # Test of improperly shaped X and Y
+
+
+if __name__ == "__main__":
+    run_tests()
```
Comment (Luthaf) [inline tests/pcovr.py]: You may want to have a look at the [unittest](https://docs.python.org/3/library/unittest.html) module which provides facilities to write such tests; and [tox](https://tox.readthedocs.io/) as a test runner to setup test environments with multiple python versions.

File: utilities/sklearn_covr/kpcovr.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+        The number of parallel jobs to run.
+        ``None`` means 1 unless in a :obj:`joblib.parallel_backend` context.
+        ``-1`` means using all processors. See :term:`Glossary <n_jobs>`
+        for more details.
+
+        .. versionadded:: 0.18
+
+    Attributes
+    ----------
+    lambdas_ : array, (n_components,)
```
Comment (Luthaf) [inline utilities/sklearn_covr/kpcovr.py:95]: This does not correspond to the full list of attributes, but I don't know what's the sklearn policy on this

File: utilities/sklearn_covr/kpcovr.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+
+    X_transformed_fit_ : array, (n_samples, n_components)
+        Projection of the fitted data on the kernel principal components.
+        Only available when ``fit_inverse_transform`` is True.
+
+    X_fit_ : (n_samples, n_features)
+        The data used to fit the model. If `copy_X=False`, then `X_fit_` is
+        a reference. This attribute is used for the calls to transform.
+
+    Examples
```
Comment (Luthaf) [inline utilities/sklearn_covr/kpcovr.py]: c/p leftover?

**Review by Luthaf — COMMENTED** [review]
Looks like a good starting point to me, I have not run the tests or checked that the code produces the correct results though =)

File: utilities/sklearn_covr/pcovr.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+            else:
+                lr = LR(**self.lr_params)  # some sort of args
+                lr.fit(X, Y)
+                self.Yhat = lr.predict(X)
+                self.W = lr.coef_.T
+
+        if self.W is None:
+            W = np.linalg.pinv(np.dot(X.T, X), rcond=self.regularization)
+            W = mdot([W, X.T, Y])
+            self.W = W
```
Comment (hurricane642) [inline utilities/sklearn_covr/pcovr.py]: Why we use here an LSA without a regularization parameter?

File: utilities/sklearn_covr/pcovr.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        if self.Yhat is None or self.W is None:
+            self._compute_Yhat(X, Y)
+
+        # Sparse eigensolvers will not work when seeking N-1 eigenvalues
+        if min(X.shape) <= self.n_components:
+            self.full_eig = True
+
+        self._fit(X, Y)
+
+    # todo fit_transform and fit_predict
```
Comment (hurricane642) [inline utilities/sklearn_covr/pcovr.py]:
```suggestion
    def fit_transform(self, X, Y, Yhat=None, W=None):
        #return projection in latent spase, targets and the reconstructed input data
        self.fit(X,Y, Yhat, W)
        T = self._project(X, 'pxt_')
        Yp = self._project(T, 'pty_')
        Xr = self._project(T, 'ptx_')
        return T, Yp, Xr

    def fit_predict(self, X, Y, Yhat=None, W=None):
        self.fit( X, Y, Yhat, W )
        return self.predict(X)
```

File: utilities/sklearn_covr/_base.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        Returns
+        -------
+        A' : array-like, shape (n, p)
+
+        Examples
+        --------
+
+        >>> todo
+        """
+        check_is_fitted(self)
```
Comment (hurricane642) [inline utilities/sklearn_covr/_base.py]:
```suggestion
        check_is_fitted(self, projector)
```

File: utilities/sklearn_covr/kpcovr.py (hunk 1 above)
Comment (rosecers) [reply utilities/sklearn_covr/kpcovr.py:83]: The `versionadded`? Likely. But should we keep it and change to `scikit-learn versionadded`?

File: utilities/sklearn_covr/pcovr.py (hunk 5 above)
Comment (rosecers) [reply utilities/sklearn_covr/pcovr.py]: What exactly do you mean? `rcond` is the the regularization parameter

File: utilities/sklearn_covr/pcovr.py (hunk 6 above)
Comment (rosecers) [reply utilities/sklearn_covr/pcovr.py]: I would say that `fit_transform` should only return T, no? If we wanted to return `Xr` we could potentially do `fit_inverse_transform()`

File: utilities/sklearn_covr/kpcovr.py [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+    --------
+    >>> from sklearn.datasets import load_digits
+    >>> from sklearn.decomposition import KernelPCA
+    >>> X, _ = load_digits(return_X_y=True)
+    >>> transformer = KernelPCA(n_components=7, kernel='linear')
+    >>> X_transformed = transformer.fit_transform(X)
+    >>> X_transformed.shape
+    (1797, 7)
```
Comment (rosecers) [inline utilities/sklearn_covr/kpcovr.py]:
```suggestion
```

File: utilities/sklearn_covr/pcovr.py (hunk 6 above)
Comment (hurricane642) [reply utilities/sklearn_covr/pcovr.py]: Make sense

Comment (hurricane642) [reply utilities/sklearn_covr/pcovr.py]:
```suggestion
def fit_transform(self, X, Y, Yhat=None, W=None):
        #return projection in latent spase, targets and the reconstructed input data
        self.fit(X,Y, Yhat, W)
        T = self._project(X, 'pxt_')
        return T
def fit_predict(self, X, Y, Yhat=None, W=None):
        self.fit( X, Y, Yhat, W )
        return self.predict(X)
```

File: utilities/sklearn_covr/pcovr.py (hunk 5 above)
Comment (hurricane642) [reply utilities/sklearn_covr/pcovr.py]: Aaaa, Ok, i thought, that it only the rcut for the small singular values.

File: tests/pcovr.py (hunk 2 above)
Comment (rosecers) [reply tests/pcovr.py]: @Luthaf done! lmk thoughts

File: tests/_base.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                                )
+
+    # Checks that the model will not transform before fitting
+    def test_nonfitted_failure(self):
+        model = self.model(mixing=0.5,
+                           n_components=2,
+                           tol=1E-12)
+        with self.assertRaises(exceptions.NotFittedError):
+            if(self.K is not None):
+                _ = model.transform(self.X, K=self.X)
```
Comment (hurricane642) [inline tests/_base.py]:
```suggestion
                _ = model.transform(self.X, K=self.K)
```

File: tests/_base.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+        for i, mixing in enumerate(self.alphas):
+            if(np.isnan(self.lr_errors[i]) or np.isnan(self.pca_errors[i])):
+                Yp, _, Xr = self.run_pcovr(mixing=mixing)
+                self.lr_errors[i] = self.rel_error(self.Y, Yp)
+                self.pca_errors[i] = self.rel_error(self.X, Xr)
+
+    # Basic Test of model LR Errors, that None return np.nan
+
+    def test_lr_errors(self):
+        self.add_errors()
```
Comment (hurricane642) [inline tests/_base.py:61]: After this function, all lr_errors become not None, or I didn't understand something?

File: tests/_base.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+    # Test that model PCA Errors are monotonic with alpha
+    def test_pca_monotonicity(self):
+        self.add_errors()
+
+        for i, a in enumerate(self.alphas[:-1]):
+            with self.subTest(i=i):
+                pca1 = round(self.pca_errors[i], self.rounding)
+                pca2 = round(self.pca_errors[i+1], self.rounding)
+                self.assertTrue(pca1 >= pca2,
+                                msg=f'PCA Error Non-Monotonic\n {pca1} >  {pca2}'
```
Comment (hurricane642) [inline tests/_base.py]: vice versa, in case of error pca1<pca2

File: tests/pcovr.py (hunk 2 above)
Comment (Luthaf) [reply tests/pcovr.py]: Tests look better, I don't have much time right now to look at the code =)

File: tests/_base.py (hunk 10 above)
Comment (rosecers) [reply tests/_base.py:61]: The point is to run the model once per alpha and then use the results in multiple tests.

File: tests/_base.py (hunk 11 above)
Comment (rosecers) [inline tests/_base.py]:
```suggestion
                                msg=f'PCA Error Non-Monotonic\n {pca1} < {pca2}'
```

**Comment by rosecers** [comment]
@hurricane642 can you a PR on sklearn-cosmo for this? We'll delete this one here once it's been moved over

**Comment by hurricane642** [comment]
Ok, yes, I will. I'm working on it, sorry for the delay.

**Pull request closed without merging**